### PR TITLE
Blog on new memcpy APIs.

### PR DESCRIPTION
**Pull request by marcinslusarz**
This change is [Reviewable](https://reviewable.io/reviews/pmem/pmem.github.io/92)

### CODE REVIEW

**Comment by marcinslusarz** [comment]
Preview: http://marcinslusarz.github.io/2018/12/06/extended-memcpy.html

File: _posts/2018-12-06-extended-memcpy.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+(configurable using `PMEM_MOVNT_THRESHOLD` environment variable) they use
+normal `mov` instrunctions followed by `pmem_flush`, but for modifications
+256 bytes and longer they use non-temporal stores. Those instructions on
+x86_64 have 2 properties - they bypass CPU caches (so that pmem_flush is not
+needed) and treat destination memory as the write-combining type.
+The last property means that if the destination memory is not in the cache,
+the CPU doesn't have to fetch the full cache lines, just to flush them a moment
+later. This is important not because data is not stored in the cache, but
+because there's no fetch of previous data.
+This means that application can update the same piece of memory multiple times
```
Comment (pbalcer) [inline _posts/2018-12-06-extended-memcpy.md]: `can update the same cacheline multiple times?`

"Piece of memory" is ambiguous.

File: _posts/2018-12-06-extended-memcpy.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+without waiting for it to be available for read.
+
+The logic based on store size is usually optimal for sequential and random
+workloads, but fails to chose the right method when application writes to
+the same cache lines over and over again (like in case of hot meta-data).
```
Comment (pbalcer) [inline _posts/2018-12-06-extended-memcpy.md]: like in the case of how metadata.

Comment (pbalcer) [inline _posts/2018-12-06-extended-memcpy.md]: constantly writes to the same cachelines

File: _posts/2018-12-06-extended-memcpy.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+void *pmemobj_memcpy (PMEMobjpool *pop, void *dest, const void *src, size_t len,
+		      unsigned flags);
+void *pmemobj_memmove(PMEMobjpool *pop, void *dest, const void *src, size_t len,
+		      unsigned flags);
+void *pmemobj_memset (PMEMobjpool *pop, void *dest, int c, size_t len,
+		      unsigned flags);
+{% endhighlight %}
+
+As you can see, we stopped with `_persist` and `_nodrain` distinction and
```
Comment (pbalcer) [inline _posts/2018-12-06-extended-memcpy.md]: we removed the...suffixes

File: _posts/2018-12-06-extended-memcpy.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+introduced the more flexible "flags" argument.
```
Comment (pbalcer) [inline _posts/2018-12-06-extended-memcpy.md]: introduced a more flexible `flags` argument.

File: _posts/2018-12-06-extended-memcpy.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+When "flags" is 0, those functions behave like `_persist` variants of older
+APIs. For libpmem functions we introduced these flags:
```
Comment (pbalcer) [inline _posts/2018-12-06-extended-memcpy.md]: APIs. \n For libpmem...

File: _posts/2018-12-06-extended-memcpy.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+  - `PMEM_F_MEM_NONTEMPORAL`
+  - `PMEM_F_MEM_TEMPORAL`
+  - `PMEM_F_MEM_WC`
+  - `PMEM_F_MEM_WB`
+
+`PMEM_F_MEM_NODRAIN` disables `pmem_drain`, just like `_nodrain` functions
+did. 
+
+`PMEM_F_MEM_NOFLUSH` disables flushing (and implies `PMEM_F_MEM_NODRAIN`).
+This flag may be useful if application knows it will update the same piece of
```
Comment (pbalcer) [inline _posts/2018-12-06-extended-memcpy.md]: change the piece of memory ;)
Maybe "region" ? "destination buffer" ?

File: _posts/2018-12-06-extended-memcpy.md [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+(so it should keep it in CPU cache). On x86_64 this means the usage of normal
+stores. On x86_64 without CLWB instruction (or motherboard without eADR)
+the cache will still be invalidated (unless `PMEM_F_MEM_NOFLUSH` was also
+used).
+
+`PMEM_F_MEM_WB` tells the library to treat destination memory as write-back.
+On x86_64 this means the usage of normal stores.
+
+Few more notes on non-temporal stores usage on x86_64:
+- If you use `PMEM_F_MEM_NODRAIN`, you should call `pmem_drain` at least
```
Comment (pbalcer) [inline _posts/2018-12-06-extended-memcpy.md]: once you want...

File: _posts/2018-12-06-extended-memcpy.md [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+  when you want the copied data to be visible to other threads, because without
```
Comment (pbalcer) [inline _posts/2018-12-06-extended-memcpy.md]: threads. Otherwise, that data may be stuck in ...

File: _posts/2018-12-06-extended-memcpy.md [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+  that your data may be stuck in (per CPU) write-combining buffers.
+- It's important to align modified data to cache-line boundary (64 bytes),
```
Comment (pbalcer) [inline _posts/2018-12-06-extended-memcpy.md]: `to align destination buffer`
?

File: _posts/2018-12-06-extended-memcpy.md [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+  - `PMEMOBJ_F_MEM_NONTEMPORAL`
+  - `PMEMOBJ_F_MEM_TEMPORAL`
+  - `PMEMOBJ_F_MEM_WC`
+  - `PMEMOBJ_F_MEM_WB`
+  - `PMEMOBJ_F_RELAXED`
+
+First 6 behave like their libpmem equivalents.
+
+`PMEMOBJ_F_RELAXED` matters only when used with remote replication.
+For now RDMA protocol (as used by librpmem to implement remote replication)
```
Comment (pbalcer) [inline _posts/2018-12-06-extended-memcpy.md]: At the present, the RDMA protocol

File: _posts/2018-12-06-extended-memcpy.md [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+doesn't guarantee that in case of interruption the transfer won't be torn
```
Comment (pbalcer) [inline _posts/2018-12-06-extended-memcpy.md]: interruption, the transfer won't be

File: _posts/2018-12-06-extended-memcpy.md [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+at random place. In order to provide 8 byte atomicity guarantee mentioned
+earlier, pmemobj have to use slower method of replication. This flag tells
```
Comment (pbalcer) [inline _posts/2018-12-06-extended-memcpy.md]: has to use a slower method

File: _posts/2018-12-06-extended-memcpy.md [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+the library that the application doesn't care about this guarantee for
+this operation, so it can replicate using faster method.
```
Comment (pbalcer) [inline _posts/2018-12-06-extended-memcpy.md]: using a faster method.

File: _posts/2018-12-06-extended-memcpy.md [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+
+If you are overwhelmed by this knowledge, sticking to older functions or setting
+flags argument to 0 for the new ones is a safe choice. However if you want
+to squeeze as much of performance from your brand new NVDIMMs as possible,
```
Comment (pbalcer) [inline _posts/2018-12-06-extended-memcpy.md]: squeeze as much performance as possible from your brand new NVDIMMs, it's may be worth using those new APIs.

File: _posts/2018-12-06-extended-memcpy.md [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+`PMEM_F_MEM_WB` tells the library to treat destination memory as write-back.
+On x86_64 this means the usage of normal stores.
+
+Few more notes on non-temporal stores usage on x86_64:
+- If you use `PMEM_F_MEM_NODRAIN`, you should call `pmem_drain` at least
+  when you want the copied data to be visible to other threads, because without
+  that your data may be stuck in (per CPU) write-combining buffers.
+- It's important to align modified data to cache-line boundary (64 bytes),
+  because write-combining happens on cache-line level.
+  In pmemobj you can use allocation classes with multiple of 64 bytes alignment
```
Comment (wlemkows) [inline _posts/2018-12-06-extended-memcpy.md]: Do we have any blog/doc about allocation classes to refer?

**Pull request merged**